modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/thunks/fieldChange.es.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {debounce} from 'frontend-js-web';
+
+import {EVENT_TYPES} from '../actions/types.es';
+import {evaluate} from '../util/evaluation.es';
+import {PagesVisitor} from '../util/visitors.es';
+
+let REVALIDATE_UPDATES = [];
+
+const getEditedPages = ({editingLanguageId, name, pages, value}) => {
+	const pageVisitor = new PagesVisitor(pages);
+
+	return pageVisitor.mapFields(
+		(field) => {
+			if (field.name === name) {
+				return {
+					...field,
+					localizedValue: {
+						...field.localizedValue,
+						[editingLanguageId]: value,
+					},
+					value,
+				};
+			}
+
+			return field;
+		},
+		false,
+		true
+	);
+};
+
+const debounceFn = debounce((fn) => fn(), 50);
+
+export default function fieldChange({
+	defaultLanguageId,
+	editingLanguageId,
+	pages,
+	portletNamespace,
+	properties,
+	rules,
+}) {
+	return (dispatch) =>
+
+		// We added an event delay to break long tasks into short
+		// durations that are done when `fieldChange` is called
+		// very often. That's because changing a fieldEdited is a
+		// very expensive operation that causes many renderings in
+		// the application.
+
+		debounceFn(() => {
+			const {fieldInstance, value} = properties;
+			const {evaluable, fieldName} = fieldInstance;
+
+			const editedPages = getEditedPages({
+				editingLanguageId,
+				name: fieldInstance.name,
+				pages,
+				value,
+			});
+
+			// We want a synchronous update without waiting for an evaluation of
+			// the field.
+
+			dispatch({payload: editedPages, type: EVENT_TYPES.UPDATE_PAGES});
+
+			// We triggered a dispatch of FIELD_CHANGE just to propagate the event to
+			// the upper layers.
+
+			dispatch({payload: properties, type: EVENT_TYPES.FIELD_CHANGE});
```

Comment:
Always a bit of a smell to dispatch two things in a row; maybe there is a single action that can encompass both of these as a single unit?

Reply:
Hmm, I think, it wouldn't be possible. Why the `UPDATE_PAGES` event always receives new pages. And we do not handle `FIELD_CHANGE` events on FormProvider, this is because of LayoutProvider which handles other operations.